src/irmin/tree.ml

```diff
@@ -600,9 +602,28 @@ module Make (P : S.PRIVATE) = struct
           match (cached_value x, cached_value y) with
           | Some x, Some y -> equal_node x y
           | _ -> (
-              match (cached_map x, cached_map y) with
-              | Some x, Some y -> map_equal x y
-              | _ -> hash_equal (hash x) (hash y)))
+              match (cached_hash x, cached_hash y) with
+              | Some x, Some y -> equal_hash x y
+              | _ -> (
+                  match (cached_map x, cached_map y) with
+                  | Some x, Some y -> map_equal x y
+                  | _ -> hash_equal (hash x) (hash y))))
+
+    (* same as [equal] but do not compare in-memory maps
+       recursively. *)
+    let maybe_equal (x : t) (y : t) =
+      x == y
+      ||
+      match (x.v, y.v) with
+      | Hash (_, x), Hash (_, y) -> hash_equal x y
+      | Value (_, x, None), Value (_, y, None) -> equal_node x y
```

Comment:
You can trop the 2 first cases as they are redundant with the ones below using `cached_value` and `cached_hash`

Reply:
it's not totally redundant: we have to decide which equality function we will be using first depending on their underlying representation. I don't know if that makes a big difference (probably not) so yes can simplify this.

Reply:
Ok I see what you mean.

I tried with and without and I couldn't see any difference.